File: astrx-suite/torrentds/torrentds/bencode.py
"""Hand-rolled bencode (BEP-3) encoder/decoder.

Supports the four bencode types: integers, byte strings, lists and
dictionaries.  Encoding is *canonical*: dictionary keys are emitted in
lexicographic (byte-wise) order and integers carry no redundant leading
zeros or negative-zero.  This module underpins KRPC (DHT), the tracker
protocols and the .torrent info-dict parsing, so it is deliberately strict.

Design notes
------------
* ``encode`` accepts ``int``, ``bytes``/``bytearray``, ``str`` (UTF-8
  encoded), ``list``/``tuple`` and ``dict``.  ``bool`` is intentionally
  rejected -- bencode has no boolean and silently coercing it hides bugs.
* ``decode`` returns ``int``, ``bytes``, ``list`` and ``dict`` (with
  ``bytes`` keys).  Strings are never auto-decoded to ``str`` because
  info-hashes and peer data are binary.
* ``decode`` is strict: trailing garbage, leading zeros, ``-0``, unsorted
  or duplicate dict keys and truncated input all raise ``BencodeError``.
  This strictness is what lets the DHT re-hash an info dict and trust the
  result.
"""

from __future__ import annotations

from typing import Any, Tuple

__all__ = ["encode", "decode", "decode_prefix", "decode_lenient", "BencodeError"]


class BencodeError(ValueError):
    """Raised on any malformed bencode input or unencodable object."""


# Maximum container nesting accepted by ``decode``.  Adversarial input can
# nest lists/dicts arbitrarily; without a bound the recursive decoder hits
# Python's recursion limit and raises ``RecursionError`` (which is *not* a
# ``ValueError``, so it would escape callers).  100 is far deeper than any
# real KRPC message or info-dict.
MAX_DEPTH = 100


# --------------------------------------------------------------------------
# Encoding
# --------------------------------------------------------------------------

def encode(obj: Any) -> bytes:
    """Serialise *obj* to canonical bencode bytes."""
    out: list[bytes] = []
    _encode(obj, out)
    return b"".join(out)


def _encode(obj: Any, out: list[bytes]) -> None:
    # bool must be checked before int (bool is a subclass of int).
    if isinstance(obj, bool):
        raise BencodeError("bencode has no boolean type")
    if isinstance(obj, int):
        out.append(b"i%de" % obj)
    elif isinstance(obj, (bytes, bytearray)):
        out.append(b"%d:" % len(obj))
        out.append(bytes(obj))
    elif isinstance(obj, str):
        data = obj.encode("utf-8")
        out.append(b"%d:" % len(data))
        out.append(data)
    elif isinstance(obj, (list, tuple)):
        out.append(b"l")
        for item in obj:
            _encode(item, out)
        out.append(b"e")
    elif isinstance(obj, dict):
        out.append(b"d")
        items = []
        for key, value in obj.items():
            if isinstance(key, str):
                key = key.encode("utf-8")
            elif isinstance(key, (bytes, bytearray)):
                key = bytes(key)
            else:
                raise BencodeError("dict keys must be bytes or str")
            items.append((key, value))
        # Canonical form: keys sorted by raw byte value; duplicates illegal.
        items.sort(key=lambda kv: kv[0])
        for i in range(1, len(items)):
            if items[i][0] == items[i - 1][0]:
                raise BencodeError("duplicate dict key: %r" % items[i][0])
        for key, value in items:
            out.append(b"%d:" % len(key))
            out.append(key)
            _encode(value, out)
        out.append(b"e")
    else:
        raise BencodeError("cannot bencode object of type %s" % type(obj).__name__)


# --------------------------------------------------------------------------
# Decoding
# --------------------------------------------------------------------------

def decode(data: bytes) -> Any:
    """Decode a complete bencode byte string.

    Raises ``BencodeError`` if *data* is malformed or has trailing bytes.
    """
    if not isinstance(data, (bytes, bytearray)):
        raise BencodeError("decode expects bytes")
    data = bytes(data)
    try:
        value, index = _decode(data, 0)
    except RecursionError as exc:
        raise BencodeError("bencode nested too deeply") from exc
    except (IndexError, ValueError) as exc:
        raise BencodeError("truncated or invalid bencode: %s" % exc) from exc
    if index != len(data):
        raise BencodeError("trailing bytes after bencode value")
    return value


def decode_prefix(data: bytes) -> Tuple[Any, int]:
    """Decode one bencode value from the front of *data*.

    Returns ``(value, bytes_consumed)`` and permits trailing bytes.  Used
    by the ut_metadata (BEP-9) *data* message, which appends raw piece
    bytes immediately after a bencoded header dict.
    """
    if not isinstance(data, (bytes, bytearray)):
        raise BencodeError("decode_prefix expects bytes")
    try:
        value, index = _decode(bytes(data), 0)
    except RecursionError as exc:
        raise BencodeError("bencode nested too deeply") from exc
    except (IndexError, ValueError) as exc:
        raise BencodeError("truncated or invalid bencode: %s" % exc) from exc
    return value, index


def _decode(data: bytes, index: int, depth: int = 0) -> Tuple[Any, int]:
    if depth > MAX_DEPTH:
        raise BencodeError("bencode nested too deeply (>%d)" % MAX_DEPTH)
    if index >= len(data):
        raise BencodeError("unexpected end of data")
    ch = data[index : index + 1]
    if ch == b"i":
        return _decode_int(data, index)
    if ch == b"l":
        return _decode_list(data, index, depth)
    if ch == b"d":
        return _decode_dict(data, index, depth)
    if ch.isdigit():
        return _decode_bytes(data, index)
    raise BencodeError("invalid token %r at position %d" % (ch, index))


def _decode_int(data: bytes, index: int) -> Tuple[int, int]:
    end = data.index(b"e", index)
    body = data[index + 1 : end]
    if body in (b"", b"-"):
        raise BencodeError("empty integer")
    neg = body.startswith(b"-")
    digits = body[1:] if neg else body
    if not digits.isdigit():
        raise BencodeError("non-numeric integer: %r" % body)
    # Reject non-canonical encodings: -0, and leading zeros like 03 / -05.
    if neg and digits == b"0":
        raise BencodeError("negative zero is not canonical")
    if len(digits) > 1 and digits[0:1] == b"0":
        raise BencodeError("leading zero is not canonical: %r" % body)
    return int(body), end + 1


def _decode_bytes(data: bytes, index: int) -> Tuple[bytes, int]:
    colon = data.index(b":", index)
    length_field = data[index:colon]
    if len(length_field) > 1 and length_field[0:1] == b"0":
        raise BencodeError("leading zero in string length: %r" % length_field)
    if not length_field.isdigit():
        raise BencodeError("invalid string length: %r" % length_field)
    length = int(length_field)
    start = colon + 1
    end = start + length
    if end > len(data):
        raise BencodeError("string longer than remaining data")
    return data[start:end], end


def _decode_list(data: bytes, index: int, depth: int = 0) -> Tuple[list, int]:
    result: list = []
    index += 1  # skip 'l'
    while True:
        if index >= len(data):
            raise BencodeError("unterminated list")
        if data[index : index + 1] == b"e":
            return result, index + 1
        value, index = _decode(data, index, depth + 1)
        result.append(value)


def _decode_dict(data: bytes, index: int, depth: int = 0) -> Tuple[dict, int]:
    result: dict = {}
    index += 1  # skip 'd'
    last_key: bytes | None = None
    while True:
        if index >= len(data):
            raise BencodeError("unterminated dict")
        if data[index : index + 1] == b"e":
            return result, index + 1
        if not data[index : index + 1].isdigit():
            raise BencodeError("dict key must be a byte string")
        key, index = _decode_bytes(data, index)
        if last_key is not None and key <= last_key:
            # Enforces both ordering and no-duplicate-keys for canonical form.
            raise BencodeError("dict keys not sorted / duplicated: %r" % key)
        last_key = key
        value, index = _decode(data, index, depth + 1)
        result[key] = value


# --------------------------------------------------------------------------
# Lenient decoding (info-dict only)
# --------------------------------------------------------------------------
#
# Real-world .torrent info-dicts produced by careless clients are sometimes
# mildly non-canonical: dict keys out of order, a duplicated key, an integer
# or string length with a redundant leading zero.  The *strict* decoder above
# rejects all of these -- which is exactly what we want for KRPC and any other
# network-facing decode -- but for the **info-dict** it just costs us fetch
# yield, because the info-dict has already been SHA-1-verified against the
# infohash on its RAW bytes (see ``metadata.assemble_and_verify``).  Relaxing
# canonical-form checks here therefore never weakens the ``sha1(info) ==
# infohash`` guarantee: verification happens on the untouched wire bytes
# *before* we ever call this.
#
# This path is used ONLY for the metadata info-dict.  It is never wired into
# ``parse_message`` / KRPC or any other decode -- those keep the strict
# decoder.  All the memory-safety bounds (MAX_DEPTH, length-vs-remaining
# checks) are preserved; only the canonical-form checks are dropped.

def decode_lenient(data: bytes) -> Any:
    """Tolerantly decode a bencode value (unsorted/dup keys, leading zeros).

    Intended solely for SHA-1-verified info-dict bytes.  Still requires the
    whole buffer to be consumed and still enforces :data:`MAX_DEPTH`.
    """
    if not isinstance(data, (bytes, bytearray)):
        raise BencodeError("decode_lenient expects bytes")
    data = bytes(data)
    try:
        value, index = _decode_lenient(data, 0)
    except RecursionError as exc:
        raise BencodeError("bencode nested too deeply") from exc
    except (IndexError, ValueError) as exc:
        raise BencodeError("truncated or invalid bencode: %s" % exc) from exc
    if index != len(data):
        raise BencodeError("trailing bytes after bencode value")
    return value


def _decode_lenient(data: bytes, index: int, depth: int = 0) -> Tuple[Any, int]:
    if depth > MAX_DEPTH:
        raise BencodeError("bencode nested too deeply (>%d)" % MAX_DEPTH)
    if index >= len(data):
        raise BencodeError("unexpected end of data")
    ch = data[index : index + 1]
    if ch == b"i":
        return _decode_int_lenient(data, index)
    if ch == b"l":
        return _decode_list_lenient(data, index, depth)
    if ch == b"d":
        return _decode_dict_lenient(data, index, depth)
    if ch.isdigit():
        return _decode_bytes_lenient(data, index)
    raise BencodeError("invalid token %r at position %d" % (ch, index))


def _decode_int_lenient(data: bytes, index: int) -> Tuple[int, int]:
    end = data.index(b"e", index)
    body = data[index + 1 : end]
    if body in (b"", b"-"):
        raise BencodeError("empty integer")
    digits = body[1:] if body.startswith(b"-") else body
    if not digits.isdigit():
        raise BencodeError("non-numeric integer: %r" % body)
    # Leading zeros / -0 tolerated here (canonical-form only, not a safety issue).
    return int(body), end + 1


def _decode_bytes_lenient(data: bytes, index: int) -> Tuple[bytes, int]:
    colon = data.index(b":", index)
    length_field = data[index:colon]
    if not length_field.isdigit():
        raise BencodeError("invalid string length: %r" % length_field)
    length = int(length_field)   # leading-zero length tolerated
    start = colon + 1
    end = start + length
    if end > len(data):
        raise BencodeError("string longer than remaining data")
    return data[start:end], end


def _decode_list_lenient(data: bytes, index: int, depth: int = 0) -> Tuple[list, int]:
    result: list = []
    index += 1  # skip 'l'
    while True:
        if index >= len(data):
            raise BencodeError("unterminated list")
        if data[index : index + 1] == b"e":
            return result, index + 1
        value, index = _decode_lenient(data, index, depth + 1)
        result.append(value)


def _decode_dict_lenient(data: bytes, index: int, depth: int = 0) -> Tuple[dict, int]:
    result: dict = {}
    index += 1  # skip 'd'
    while True:
        if index >= len(data):
            raise BencodeError("unterminated dict")
        if data[index : index + 1] == b"e":
            return result, index + 1
        if not data[index : index + 1].isdigit():
            raise BencodeError("dict key must be a byte string")
        key, index = _decode_bytes_lenient(data, index)
        value, index = _decode_lenient(data, index, depth + 1)
        # Out-of-order keys accepted; a duplicated key keeps the last value.
        result[key] = value

File: astrx-suite/crates/torrentds/src/bencode.rs
//! Canonical bencode codec.
//!
//! The STRICT decoder rejects every non-canonical form (leading zeros, negative
//! zero, unsorted/duplicate dict keys, trailing bytes) and is bounded against
//! hostile input (depth cap, bounds checks) — it returns `Result`, never panics
//! on adversarial bytes. Dict canonicality is enforced *by construction*: keys
//! live in a `BTreeMap`, so `encode` cannot emit a non-canonical dict.

use std::collections::BTreeMap;

/// A decoded bencode value. Dict keys are raw bytes, kept sorted by `BTreeMap`.
#[derive(Debug, Clone, PartialEq, Eq)]
pub enum Ben {
    /// Bencode integer. Bounded to `i64`: every real BitTorrent/KRPC integer
    /// (file sizes, ports, timestamps, counts) fits, and rejecting an
    /// out-of-`i64` integer as malformed is the safe direction. (Python's
    /// arbitrary-precision `int` accepts larger values; that is a language
    /// artifact, not a protocol requirement — see `rejects_out_of_range_int`.)
    Int(i64),
    Bytes(Vec<u8>),
    List(Vec<Ben>),
    Dict(BTreeMap<Vec<u8>, Ben>),
}

/// Every malformed-input / unencodable error (like Python's `BencodeError`).
#[derive(Debug, PartialEq, Eq)]
pub struct BencodeError(pub String);

impl std::fmt::Display for BencodeError {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        write!(f, "bencode: {}", self.0)
    }
}
impl std::error::Error for BencodeError {}

fn err<T>(msg: impl Into<String>) -> Result<T, BencodeError> {
    Err(BencodeError(msg.into()))
}

/// Max container nesting; bounds recursion so adversarial input can't blow the stack.
const MAX_DEPTH: usize = 100;

/// Serialise to canonical bencode. Cannot produce a non-canonical dict.
pub fn encode(v: &Ben) -> Vec<u8> {
    let mut out = Vec::new();
    enc(v, &mut out);
    out
}

fn enc(v: &Ben, out: &mut Vec<u8>) {
    match v {
        Ben::Int(n) => out.extend_from_slice(format!("i{n}e").as_bytes()),
        Ben::Bytes(b) => {
            out.extend_from_slice(format!("{}:", b.len()).as_bytes());
            out.extend_from_slice(b);
        }
        Ben::List(items) => {
            out.push(b'l');
            items.iter().for_each(|it| enc(it, out));
            out.push(b'e');
        }
        Ben::Dict(m) => {
            out.push(b'd');
            for (k, val) in m {
                // BTreeMap iterates in sorted key order => canonical, always.
                out.extend_from_slice(format!("{}:", k.len()).as_bytes());
                out.extend_from_slice(k);
                enc(val, out);
            }
            out.push(b'e');
        }
    }
}

/// Decode one complete canonical value; trailing bytes are an error.
pub fn decode(data: &[u8]) -> Result<Ben, BencodeError> {
    let (v, i) = dec(data, 0, 0, true)?;
    if i != data.len() {
        return err("trailing bytes after value");
    }
    Ok(v)
}

/// Decode one value from the front, permitting trailing bytes; returns the value
/// and the number of bytes consumed. Strict (canonical) like [`decode`]. Used by
/// the ut_metadata (BEP-9) `data` message, which appends raw piece bytes right
/// after a bencoded header dict.
pub fn decode_prefix(data: &[u8]) -> Result<(Ben, usize), BencodeError> {
    dec(data, 0, 0, true)
}

/// Tolerantly decode one complete value: out-of-order / duplicate dict keys and
/// redundant leading zeros (and `-0`) are accepted (a duplicate key keeps the
/// last value). Depth and length bounds are still enforced.
///
/// Intended ONLY for SHA-1-verified info-dict bytes: the info-dict is hashed
/// against the infohash on its *raw* wire bytes *before* this is ever called
/// (see the metadata path), so relaxing canonical-form checks here cannot weaken
/// the `sha1(info) == infohash` guarantee. It is never wired into `parse_message`
/// / KRPC or any other network decode — those keep the strict decoder.
pub fn decode_lenient(data: &[u8]) -> Result<Ben, BencodeError> {
    let (v, i) = dec(data, 0, 0, false)?;
    if i != data.len() {
        return err("trailing bytes after value");
    }
    Ok(v)
}

fn dec(data: &[u8], i: usize, depth: usize, strict: bool) -> Result<(Ben, usize), BencodeError> {
    if depth > MAX_DEPTH {
        return err(format!("nested too deeply (>{MAX_DEPTH})"));
    }
    match data.get(i) {
        None => err("unexpected end of data"),
        Some(b'i') => dec_int(data, i, strict),
        Some(b'l') => dec_list(data, i, depth, strict),
        Some(b'd') => dec_dict(data, i, depth, strict),
        Some(c) if c.is_ascii_digit() => dec_bytes(data, i, strict),
        Some(c) => err(format!("invalid token {:?} at {i}", *c as char)),
    }
}

fn dec_int(data: &[u8], i: usize, strict: bool) -> Result<(Ben, usize), BencodeError> {
    let end = data[i..]
        .iter()
        .position(|&b| b == b'e')
        .map(|p| i + p)
        .ok_or(BencodeError("unterminated integer".into()))?;
    let body = &data[i + 1..end];
    if body.is_empty() || body == b"-" {
        return err("empty integer");
    }
    let neg = body[0] == b'-';
    let digits = if neg { &body[1..] } else { body };
    if digits.is_empty() || !digits.iter().all(u8::is_ascii_digit) {
        return err("non-numeric integer");
    }
    // Canonical-form checks (leading zero, -0) are strict-only.
    if strict {
        if neg && digits == b"0" {
            return err("negative zero is not canonical");
        }
        if digits.len() > 1 && digits[0] == b'0' {
            return err("leading zero is not canonical");
        }
    }
    let n: i64 = std::str::from_utf8(body)
        .unwrap()
        .parse()
        .map_err(|_| BencodeError("integer overflow".into()))?;
    Ok((Ben::Int(n), end + 1))
}

fn dec_bytes(data: &[u8], i: usize, strict: bool) -> Result<(Ben, usize), BencodeError> {
    let colon = data[i..]
        .iter()
        .position(|&b| b == b':')
        .map(|p| i + p)
        .ok_or(BencodeError("missing ':' in string".into()))?;
    let len_field = &data[i..colon];
    if strict && len_field.len() > 1 && len_field[0] == b'0' {
        return err("leading zero in string length");
    }
    if len_field.is_empty() || !len_field.iter().all(u8::is_ascii_digit) {
        return err("invalid string length");
    }
    let n: usize = std::str::from_utf8(len_field)
        .unwrap()
        .parse()
        .map_err(|_| BencodeError("string length overflow".into()))?;
    let start = colon + 1;
    let end = start
        .checked_add(n)
        .ok_or(BencodeError("length overflow".into()))?;
    if end > data.len() {
        return err("string longer than remaining data");
    }
    Ok((Ben::Bytes(data[start..end].to_vec()), end))
}

fn dec_list(
    data: &[u8],
    mut i: usize,
    depth: usize,
    strict: bool,
) -> Result<(Ben, usize), BencodeError> {
    let mut out = Vec::new();
    i += 1; // skip 'l'
    loop {
        match data.get(i) {
            None => return err("unterminated list"),
            Some(b'e') => return Ok((Ben::List(out), i + 1)),
            _ => {
                let (v, next) = dec(data, i, depth + 1, strict)?;
                out.push(v);
                i = next;
            }
        }
    }
}

fn dec_dict(
    data: &[u8],
    mut i: usize,
    depth: usize,
    strict: bool,
) -> Result<(Ben, usize), BencodeError> {
    let mut out: BTreeMap<Vec<u8>, Ben> = BTreeMap::new();
    i += 1; // skip 'd'
    let mut last_key: Option<Vec<u8>> = None;
    loop {
        match data.get(i) {
            None => return err("unterminated dict"),
            Some(b'e') => return Ok((Ben::Dict(out), i + 1)),
            Some(c) if !c.is_ascii_digit() => return err("dict key must be a byte string"),
            _ => {
                let (Ben::Bytes(key), next) = dec_bytes(data, i, strict)? else {
                    unreachable!("dec_bytes yields Ben::Bytes")
                };
                // Ordering / no-duplicate-keys is a canonical-form (strict) check;
                // lenient decode keeps the last value for a duplicated key.
                if strict {
                    if let Some(lk) = &last_key {
                        if key <= *lk {
                            return err("dict keys not sorted / duplicated");
                        }
                    }
                }
                let (v, next2) = dec(data, next, depth + 1, strict)?;
                out.insert(key.clone(), v);
                last_key = Some(key);
                i = next2;
            }
        }
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    fn hexs(b: &[u8]) -> String {
        b.iter().map(|x| format!("{x:02x}")).collect()
    }

    // Golden values from the real Python torrentds.bencode (the spec).
    const GOLDEN_MIXED_HEX: &str =
        "6c343a7370616d693432656c313a786564313a61693165313a626c693265693365656565";

    #[test]
    fn mixed_encode_and_roundtrip() {
        let mut m = BTreeMap::new();
        m.insert(b"a".to_vec(), Ben::Int(1));
        m.insert(b"b".to_vec(), Ben::List(vec![Ben::Int(2), Ben::Int(3)]));
        let val = Ben::List(vec![
            Ben::Bytes(b"spam".to_vec()),
            Ben::Int(42),
            Ben::List(vec![Ben::Bytes(b"x".to_vec())]),
            Ben::Dict(m),
        ]);
        let enc = encode(&val);
        assert_eq!(hexs(&enc), GOLDEN_MIXED_HEX);
        assert_eq!(encode(&decode(&enc).unwrap()), enc); // re-encode is canonical
    }

    #[test]
    fn rejects_malformed() {
        for bad in [
            &b"i03e"[..],
            b"i-0e",
            b"i1ejunk",
            b"d1:bi1e1:ai2ee",
            b"01:a",
            b"5:ab",
            b"x",
            b"",
        ] {
            assert!(decode(bad).is_err(), "should reject {bad:?}");
        }
    }

    #[test]
    fn deep_nesting_is_bounded() {
        let mut d = vec![b'l'; 5000];
        d.extend(vec![b'e'; 5000]);
        assert!(decode(&d).is_err());
    }

    #[test]
    fn rejects_out_of_range_int() {
        // i64 bounds accept; one past either bound is rejected as malformed
        // (deliberate — Python's arbitrary-precision int would accept these).
        assert_eq!(decode(b"i9223372036854775807e"), Ok(Ben::Int(i64::MAX)));
        assert_eq!(decode(b"i-9223372036854775808e"), Ok(Ben::Int(i64::MIN)));
        assert!(decode(b"i9223372036854775808e").is_err());
        assert!(decode(b"i-9223372036854775809e").is_err());
        assert!(decode(b"i123456789012345678901234567890e").is_err());
    }

    #[test]
    fn decode_prefix_returns_consumed_and_allows_trailer() {
        // one value off the front, trailing bytes permitted (BEP-9 data message)
        assert_eq!(decode_prefix(b"i42eEXTRA"), Ok((Ben::Int(42), 4)));
        let (v, used) = decode_prefix(b"d1:ai1eeRAWPIECEBYTES").unwrap();
        assert_eq!(used, 8);
        assert!(matches!(v, Ben::Dict(_)));
        // still strict about canonical form on the prefix itself
        assert!(decode_prefix(b"i03e").is_err());
        // a bare complete value consumes exactly its length
        assert_eq!(
            decode_prefix(b"4:spam"),
            Ok((Ben::Bytes(b"spam".to_vec()), 6))
        );
    }

    #[test]
    fn decode_lenient_tolerates_noncanonical() {
        // unsorted keys: strict rejects, lenient accepts (last value on dup wins)
        assert!(decode(b"d1:bi1e1:ai2ee").is_err());
        let v = decode_lenient(b"d1:bi1e1:ai2ee").unwrap();
        let Ben::Dict(m) = v else { panic!("dict") };
        assert_eq!(m.get(b"a".as_slice()), Some(&Ben::Int(2)));
        assert_eq!(m.get(b"b".as_slice()), Some(&Ben::Int(1)));
        // leading zeros / -0 tolerated
        assert_eq!(decode_lenient(b"i03e"), Ok(Ben::Int(3)));
        assert_eq!(decode_lenient(b"i-0e"), Ok(Ben::Int(0)));
        assert_eq!(decode_lenient(b"03:abc"), Ok(Ben::Bytes(b"abc".to_vec())));
        // duplicate key: last wins
        let Ben::Dict(m2) = decode_lenient(b"d1:ai1e1:ai9ee").unwrap() else {
            panic!("dict")
        };
        assert_eq!(m2.get(b"a".as_slice()), Some(&Ben::Int(9)));
        // safety bounds still enforced: trailing garbage and overlong strings rejected
        assert!(decode_lenient(b"i1ejunk").is_err());
        assert!(decode_lenient(b"5:ab").is_err());
    }
}
